### PR TITLE
#341 add text_navigation remove work_selection

### PR DESCRIPTION
**Pull request by daanvdh**

### CODE REVIEW

**Comment by knausj85** [comment]
@rntz @dwiel @fidgetingbits -

I'll be merging this this later in a couple days if there are no concerns.

**Comment by rntz** [comment]
I probably won't have time to look at this till Saturday.

**Comment by Offpics** [comment]
This code doesn't work for me, it always selects a whole line. I'm on windows 10, before diving into debugging the code do I need some prereqs like a beta version?

**Comment by knausj85** [comment]
@Offpics in what application are you seeing this? I know it in Visual Studio code, you need to toggle the setting off for the moment
setting

**Comment by Offpics** [comment]
@knausj85 after this setting it works, thanks

**Review by rntz — COMMENTED** [review]
It's late and I got distracted by other things, so in particular I didn't get a chance to test this code, hoping to revisit it tomorrow. Apart from clear versus delete I didn't see anything immediately pressing, although (this is a nitpick): is up distinct from left, or down distinct from right? If not, perhaps better to have only left and right for simplicity and consistency.

I also feel like some of the commands feel a bit too verbose (couldn't we omit "symbol"?), but I haven't actually used them yet so take that with a grain of salt...

**Comment by knausj85** [comment]
yeah, I went pretty verbose by default due to a number of conflicts I encountered during initial testing. I figure we can refine these over time. 

the other thing to consider is that we may be able to reuse the more verbose commands for dictation mode, while maintaining other, shorter commands for specific editors that support similar stuff (ie metago)

**Comment by knausj85** [comment]
and yes the code supports scanning lines above and below based on a setting, so up and down are relevant

**Comment by rntz** [comment]
Okay, I've had a chance to test this now. Mostly looks good, I only have one concern before merging: Many of the commands, for example `select after symbol comma`, actually select/cut/delete/etc single words, which wasn't what I was expecting (I was expecting it to operate on everything between the cursor and the selected symbol). This is fine, but they should be named accordingly, for example `select word after symbol comma`. This is nicely forward-compatible with other commands that we could add later, for example `select three words after symbol comma`. Does that seem sensible?

I have a few other ideas that are actionable immediately but shouldn't block merging the PR:

1. Currently the commands do not work in emacs, which is fine, except that overriding them to get them to work in emacs is currently kind of a hassle: you have to override three separate actions. It should be possible to unify these so there's only one action that you actually need to override, and the other actions' implementations just call that.

2. I really do feel that "symbol" is redundant, and it should just be `select word after comma`. Does this conflict with something?

I also think it should be possible to make the code a little simpler, and have ideas for new commands that we could implement using this framework, but that can all definitely wait until the PR is merged :).

Anyway, I'm planning on working on this tonight. Will report back if I finish anything. This is a very cool PR, will be excited to see it merged!

**Comment by rntz** [comment]
uh-oh... I do have one concern that I've only just noticed. This really does a number on my DFA compilation times. Causes them to spike from about .6s to 1-1.2s. I'll try talking with aegis about it.

I don't suppose either of you have noticed an increase in DFA compilation times with this code?

**Comment by knausj85** [comment]
We're both using Dragon, so no.

**Comment by rntz** [comment]
ok, after some testing, it seems like the version that uses fewer, more-generic commands (line 69 onwards in text_navigation.talon, https://github.com/knausj85/knausj_talon/pull/366/files#diff-500614c2dbae29f38fef716fbc3d9eea35d9e81287431a0311a6404a77de6218R69) performs considerably better under DFA compilation. I talked with aegis and he seemed to prefer that version as well (slack link: https://talonvoice.slack.com/archives/G9YTMSZ2T/p1614550708479100).

What's the motivation for the current approach?

**Comment by knausj85** [comment]
The motivation doesn't really matter if it's unworkable under w2l.

The capture-based commands will similarly need to be made a bit more verbose to prevent many conflicts ime, but that's fine.

**Comment by rntz** [comment]
It would be good to get some testing on this from somebody other than me who uses wav2letter (or who is willing to try using it for a bit), my cpu seems to be particularly bad about DFA compilation.

**Comment by knausj85** [comment]
I can certainly do some testing on w2l, but mine is fairly beefy. How do we enable the DFA timing information?

**Comment by rntz** [comment]
If you have debugging enabled in engines.py I think it just shows up whenever you trigger a DFA compilation, which should be whenever you switch between contexts that have different command sets defined. For example, my standard test is just to `focus` back and forth between two different windows, ideally one of which has lots of commands defined - try your favorite editor editing a talon file.

Look for a log line that looks like:
```
2021-02-28 23:42:19    IO COMPILING
load=2.94382ms build=42.9442ms link=4.99813ms flatten=4.19855ms minimize=257.689ms cfg=3.06246ms
2021-02-28 23:42:19    IO grammar -> dfa -> cfg in 0.327510s
```

**Comment by knausj85** [comment]
Testing with VSCode, which has probably the most commands of anything in the repo.

The difference is pretty negligible:
Daan's version: `2021-02-28 16:55:34    IO grammar -> dfa -> cfg in 0.115564s`
Mine: `I2021-02-28 16:58:07    IO grammar -> dfa -> cfg in 0.137650s`

**Comment by knausj85** [comment]
Any chance you could try with VSCode, just for giggles?

Regardless, we need to design for the worst case, which is officially rntz. I'll acquiesce on this.

**Comment by mankoff** [comment]
I can provide feedback and timings tomorrow - in < 24 hours.

**Comment by rntz** [comment]
I had a first stab at reworking the generic commands to be less ambiguous and more explicit. I've also tried to list some examples of each command, I find it makes it easier to keep track of the syntax. Take a look at the top of https://github.com/rntz/knausj_talon/blob/feature/text_navigation/text/text_navigation.talon for my work so far.

The same branch also has a commit which makes it easier to override this stuff (eg. get it working in emacs) by routing everything through `navigation_regex`: https://github.com/rntz/knausj_talon/commit/8e439a84f2a7e63526cc2c420f4c8a054929d4a4#diff-982d9e3eb996f559e633f4d194def3761d909f5a3b647d1a851fead67c32c9d1

For full comparison, see: https://github.com/daanvdh/knausj_talon/compare/feature/text_navigation...rntz:feature/text_navigation

File: text/text_navigation.talon [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# symbol navigation
+move [{user.arrow_key}] symbol <user.any_alphanumeric_key> [<number_small>]: user.navigation("GO", arrow_key or "RIGHT", "DEFAULT", any_alphanumeric_key, number_small or  1)
+move [{user.arrow_key}] before symbol <user.any_alphanumeric_key> [<number_small>]: user.navigation("GO", arrow_key or "RIGHT", 'BEFORE', any_alphanumeric_key, number_small or  1)
+move [{user.arrow_key}] after symbol <user.any_alphanumeric_key> [<number_small>]: user.navigation("GO", arrow_key or "RIGHT", 'AFTER', any_alphanumeric_key, number_small or  1)
+delete [{user.arrow_key}] symbol <user.any_alphanumeric_key> [<number_small>]: user.navigation("DELETE", arrow_key or "RIGHT","DEFAULT", any_alphanumeric_key, number_small or  1)
```
[the remark this answers is not in the document]
Comment (daanvdh) [reply text/text_navigation.talon:5]: "Up" searches from the line above it until the 10th line above it. "left" searches from the current cursor position until the beginning of the line. 

I can change "delete" to "clear". Does @knausj85 also agree?

**Comment by daanvdh** [comment]
@rntz: "cutting single symbols all that useful?" Not really, but is it worth the hassle trying to remove it?

**Comment by daanvdh** [comment]
@rntz: DUPLICATE?: "select after word two" vs "select right word three"
Yes these are duplicates, I think it's worth having both.

**Comment by daanvdh** [comment]
@rntz if you want to make it easy to override only a single method, I would propose creating a method that only accept regex, this would prevent having to duplicate line 85 in your repo. We can still keep your method that accepts both strings and regex for convenience.

**Comment by daanvdh** [comment]
@rntz can you create a PR with your changes to my branch once it's cleaned up?
@knausj85 Can you confirm that we don't want to write out the full commands anymore?

**Comment by knausj85** [comment]
@daanvdh - We won't be writing out the full commands due to the DFA times on w2l. There were a few reasons I preferred doing so, but it doesn't matter anymore. You win this round! :-p

File: text/text_navigation.talon (hunk 1 above)
[the remark this answers is not in the document]
Comment (knausj85) [reply text/text_navigation.talon:5]: I agree - clear is also more w2l-friendly ime

**Comment by mankoff** [comment]
I don't have vscode but DFA compilation w/o this code is 0.3 to 0.4s on my machine - T480 laptop w/ Kubuntu. Let me know if you need any other stats/info. With this pull request loaded, it is 0.3 to 0.5s.

**Comment by knausj85** [comment]
I think we should take a moment to make these things actions

```
def get_text_left():
def get_text_right():
def get_text_up():
def get_text_down():
```

... for the purposes of supporting these commands well in the draft UI.

**Comment by daanvdh** [comment]
@rntz Any progress on the PR?

**Comment by rntz** [comment]
Sorry, had rather a busy Sunday, working on it now, should push soon.

**Comment by daanvdh** [comment]
@knausj85 Both me and @rntz made the changes as discussed last weekend. All open points are resolved.

**Comment by daanvdh** [comment]
We don't need to re-open the discussion, but after using the ordinals I find saying "select left second word" more awkward then "select left word two".

**Comment by knausj85** [comment]
I tend to agree re: ordinals, but it's a good enough start point. We can always tweak the defaults later.

**Pull request merged**